juju/juju
# [JUJU-677] Fix issue with setting a 'complex' string controller config value

The flags parsing for config key values had an option to try and coerce strings into an assumed underlying type. This totally breaks setting complex string values like snap store assertions, and also is unnecessary, as  we do the coercion server side anyway. And for deploying charms, we actually turn off client side coercion for charm config key values.

This PR removes the client side config coercion completely. The one concession for controller config, where there are 2 attributes which are lists of strings. The existing schema list type does not handle this case (it probably should), so we so convert a string like "[a,b,c]" to a list for setting the relevant controller config (feature flags and audit log exclude methods).

## QA steps

First test that the issue in the bug report is fixed, ie
`$ juju model-config snap-store-assertions="$(cat /tmp/store-assertions)"`

Then test the scenarios which use config key value parsing:
- juju add-model
- juju bootstrap
- juju controller-config
- juju model-config
- juju model-default

In all cases, pick a bool and int attribute (eg bootstrap-timeout, set-num-control-policy for bootstrap; auditing-enabled, audit-log-max-backups for controller config; num-provision-workers, automatically-retry-hooks for model config/add-model).

Check that the above config can be set (eg using "true" for bool values) and complains when a bad value that cannot be coerced is used.

Test controller config with "features" to ensure that lists work, eg
`juju controller-config features=["foo", "bar"]`


## Bug reference

https://bugs.launchpad.net/juju/+bug/1961083
